examples/advanced/llm_hf/README.md

```diff
@@ -0,0 +1,137 @@
+```
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-sft --mode 0
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-peft --mode 1
+```
+### Adaptation Step 1: iterative training
+To adapt the centralized training script to federated application, under ``launch_once'' setting, we first need to "break" the single call to `trainer.train()` into iterative calls, one for each round of training.
+For this purpose, we provided `utils/hf_sft_peft_iter.py` as an example, which is a modified version of `utils/hf_sft_peft.py`.
+Their differences are highlighted below:
+
+![diff](./figs/diff_1.png)
+![diff](./figs/diff_2.png)
+
+Note that the `trainer.train()` call is replaced by a `for` loop. At the beginning of each round (epoch in this case), we intentionally load a fixed model weights saved at the beginning, over-writing the previous round's saved model weights, then call `trainer.train(resume_from_checkpoint=True)` with `trainer.args.num_train_epochs` incremented by 1 so that previous logging results are not overwritten. 
+
+The purpose of doing so is to tell if the intended weights are succesfully loaded at each round. Without using a fixed starting model, even if the model weights are not properly loaded, the training loss curve will still follow the one-call result, which is not what we want to see. 
+
+If the intended model weights (serving as the starting point for each round, the "global model" for FL use case) is properly loaded, then we shall observe a "zig-zag" pattern in the training loss curve. This is because the model weights are reset to the same starting point at the beginning of each round, in contrast to the one-shot centralized training, where the model weights are updated continuously, and the training loss curve should follow an overall decreasing trend.
+
+To run iterative training, we use the following command:
+``` 
+python3 ./utils/hf_sft_peft_iter.py --output_path /workspace_centralized/llama2-7b-dolly-sft-iter --mode 0
+python3 ./utils/hf_sft_peft_iter.py --output_path /workspace_centralized/llama2-7b-dolly-peft-iter --mode 1
+```
+
+The PEFT curves are shown below, blue for single call, black for iterative. We can see the "zig-zag" pattern in the iterative training loss curve.
+
+![peft](./figs/cen_peft.png)
+
+Similar patterns can be observed from the SFT curves
+
+![sft](./figs/cen_sft.png)
+
+### Adaptation Step 2: federated with NVFlare
+Once we have the iterative training script ready with "starting model" loading capability, it can be easily adapted to a NVFlare trainer by using [Client API](../../hello-world/ml-to-fl/pt/README.md).
+
+The major code modifications are for receiving and returning the global model (replacing the constant one used by iterative training), as shown below:
+
+![diff](./figs/diff_fl_1.png)
+![diff](./figs/diff_fl_2.png)
```

Comment:
It might be easier to follow in these steps:

1. Federated with NVFlare (Add flare receive before train, Add flare.send after train) (use launch_once=False, means it launch a new one everytime)
2. Since the data setup is taking a lot of time, we want to only setup once throughout the job, so change to iterative style (Add flare.is_running(), change launch_once=True)

Reply:
added some more info in readme to explain this